In [27]:
import pandas as pd
from matplotlib import pyplot as plt
import plotly.express as px
import numpy as np
# These were noted as needed in the demo docs, but I wasn't sure, so I commented them out.
# Per class demo, using the ones provided by the presenters
#import plotly.graph_objs as go
#from plotly.offline import init_notebook_mode, iplot, plot
#import plotly as py

- The source file showed as a csv in the instructions, but it was an xlsx.
- So, I looked up reading in an excel file instead over first converting it to a csv
- Then, when I tried to run read_excel, it needed openpyxl, so I pip installed that ... winner!

### #2 Use pandas.read_csv to access the provided dataset:

In [28]:
#csv = pd.read_csv('Fuel_Consumption_2000-2022.csv', header = 0)
xls = pd.read_excel('Fuel_Consumption_2000-2022.xlsx', sheet_name='in')

- Output looks good as a dataframe

In [25]:
xls

ID  YEAR   MAKE        MODEL   VEHICLE CLASS  ENGINE SIZE  \
0          1  2000  ACURA        1.6EL         COMPACT          1.6   
1          2  2000  ACURA        1.6EL         COMPACT          1.6   
2          3  2000  ACURA        3.2TL        MID-SIZE          3.2   
3          4  2000  ACURA        3.5RL        MID-SIZE          3.5   
4          5  2000  ACURA      INTEGRA      SUBCOMPACT          1.8   
...      ...   ...    ...          ...             ...          ...   
22551  22552  2022  Volvo  XC40 T5 AWD     SUV - SMALL          2.0   
22552  22553  2022  Volvo  XC60 B5 AWD     SUV - SMALL          2.0   
22553  22554  2022  Volvo  XC60 B6 AWD     SUV - SMALL          2.0   
22554  22555  2022  Volvo  XC90 T5 AWD  SUV - STANDARD          2.0   
22555  22556  2022  Volvo  XC90 T6 AWD  SUV - STANDARD          2.0   

       CYLINDERS TRANSMISSION FUEL  FUEL CONSUMPTION  COMB (mpg)  EMISSIONS  
0              4           A4    X               9.2          35        186  
1              4           M5    X               8.5          37        175  
2              6          AS5    Z              12.2          28        230  
3              6           A4    Z              13.4          25        264  
4              4           A4    X              10.0          33        198  
...          ...          ...  ...               ...         ...        ...  
22551          4          AS8    Z              10.7          30        219  
22552          4          AS8    Z              10.5          30        219  
22553          4          AS8    Z              11.0          29        232  
22554          4          AS8    Z              11.5          28        236  
22555          4          AS8    Z              12.4          26        252  

[22556 rows x 12 columns]

In [30]:
len(xls)

22556

### #3 In a Markdown cell, indicate how many rows of data are in the original source.
- There are 22556 rows in the source

### #4 To better sort and group the data, we’ll want to change some data types to ‘category’.  Using the .astype() feature change the following data columns to ‘category’ of your working data frame ‘YEAR’, ‘VEHICLE CLASS’, & ‘MAKE’

In [35]:
xls_clean = xls.astype({'YEAR': 'category', 'VEHICLE CLASS': 'category', 'MAKE': 'category'})

### #5 Put the output of a .groupby() & .describe() statement into it’s own variable from the updated data frame from STEP #4.
- Use ‘YEAR’ and ‘VEHICLE CLASS’ as your group by criteria in the above command.

In [41]:
data_display = xls_clean.groupby(['YEAR', 'VEHICLE CLASS']).describe()
data_display

ID                                     \
                            count          mean         std      min   
YEAR VEHICLE CLASS                                                     
2000 COMPACT                105.0    329.219048  221.470782      1.0   
     FULL-SIZE               28.0    241.642857  174.074955     50.0   
     MID-SIZE                77.0    347.064935  202.177250      3.0   
     MINICOMPACT             11.0    328.727273  193.007301     28.0   
     MINIVAN                 26.0    275.807692  152.079195    134.0   
...                           ...           ...         ...      ...   
2022 STATION WAGON - SMALL   19.0  22235.894737  217.971429  21624.0   
     SUBCOMPACT              79.0  21945.189873  252.686695  21621.0   
     SUV - SMALL            197.0  22081.477157  275.913854  21603.0   
     SUV - STANDARD         149.0  22044.738255  239.832470  21619.0   
     TWO-SEATER              51.0  22138.490196  276.644420  21620.0   

                                                                 ENGINE SIZE  \
                                 25%      50%       75%      max       count   
YEAR VEHICLE CLASS                                                             
2000 COMPACT                   54.00    375.0    549.00    624.0       105.0   
     FULL-SIZE                104.75    142.5    431.25    544.0        28.0   
     MID-SIZE                 136.00    390.0    492.00    632.0        77.0   
     MINICOMPACT              208.00    419.0    483.50    486.0        11.0   
     MINIVAN                  167.25    186.5    387.75    586.0        26.0   
...                              ...      ...       ...      ...         ...   
2022 STATION WAGON - SMALL  22143.50  22260.0  22345.50  22549.0        19.0   
     SUBCOMPACT             21697.50  21945.0  22232.00  22324.0        79.0   
     SUV - SMALL            21878.00  22103.0  22274.00  22554.0       197.0   
     SUV - STANDARD         21840.00  22027.0  22210.00  22556.0       149.0   
     TWO-SEATER             22016.50  22166.0  22366.50  22503.0        51.0   

                                      ... COMB (mpg)       EMISSIONS  \
                                mean  ...        75%   max     count   
YEAR VEHICLE CLASS                    ...                              
2000 COMPACT                2.348571  ...      34.00  55.0     105.0   
     FULL-SIZE              3.710714  ...      28.25  30.0      28.0   
     MID-SIZE               3.029870  ...      29.00  34.0      77.0   
     MINICOMPACT            3.263636  ...      27.50  29.0      11.0   
     MINIVAN                3.357692  ...      26.00  26.0      26.0   
...                              ...  ...        ...   ...       ...   
2022 STATION WAGON - SMALL  1.905263  ...      36.00  61.0      19.0   
     SUBCOMPACT             3.098734  ...      31.00  40.0      79.0   
     SUV - SMALL            2.328426  ...      32.00  49.0     197.0   
     SUV - STANDARD         3.970470  ...      25.00  42.0     149.0   
     TWO-SEATER             4.035294  ...      27.00  36.0      51.0   

                                                                               \
                                  mean        std    min    25%    50%    75%   
YEAR VEHICLE CLASS                                                              
2000 COMPACT                213.371429  33.433993  138.0  193.0  216.0  232.0   
     FULL-SIZE              248.535714  22.802667  218.0  229.5  246.0  260.5   
     MID-SIZE               239.000000  22.919769  189.0  223.0  235.0  260.0   
     MINICOMPACT            257.090909  21.092437  225.0  239.5  264.0  271.5   
     MINIVAN                266.076923  20.478131  246.0  253.0  261.0  274.0   
...                                ...        ...    ...    ...    ...    ...   
2022 STATION WAGON - SMALL  192.473684  39.748331  110.0  184.0  200.0  217.0   
     SUBCOMPACT             248.341772  51.964704  165.0  211.0  242.0  286.0   
 

- Misc presentation detail and for explanation in class

In [45]:
#data_display.index

In [46]:
#data_display.dtypes

In [ ]:
#[x[0] for x in data_display.index]

In [ ]:
#[x[1] for x in data_display.index]

In [53]:
#data_display['COMB (mpg)']['mean'].values.tolist()

### #6 With the variable created in STEP #5, in a new coding cell, use Plotly to create a line graph showing the mean of  'COMB (mpg)’ of each ‘VEHICLE CLASS’ across ‘YEAR’.
- Label the axes ‘YEAR’ (x-axis) and ‘Combine MPG Average’ (y-axis) [Please note, the x- and y-axis assignment was corrected on 2/27/23]
- Set the markers value as ‘True’ (line markers)
- Title the figure ‘Average Combine MPG per Vehicle Class per Year’

In [54]:
Year = [x[0] for x in data_display.index]
combineMPG = data_display['COMB (mpg)']['mean'].values.tolist()
vehicleClass = [x[1] for x in data_display.index]

In [64]:
fig = px.line(x=Year,
              y=combineMPG,
              color=vehicleClass,
              markers=True
             )

fig.update_layout(title='Average Combined MPG per Vhicle Class per Year',
                  xaxis_title='Year',
                  yaxis_title='Combined MPG Average',
                  legend_title='Vehicle Class'
                 )

fig.show()

### #7 In a Markdown cell talk about what interactive features in Plotly you find most appealing and why.
- The dyanmic functionality that shows interactive datapoints enables quick viewing of specific locations
- It is also nice to be able to remove clutter as desired by clicking on the items in the legend
